Release v1.0

XDanielPaul opened the pull request:
This MR contains v1.0 release of idf-examples-launchpad-ci-action.

It's purpose is to automate the process of building merged binaries of ESP examples using the idf-build-apps tool and a Python script that merges the binaries and generates a config.toml file. These binaries and config.toml are then uploaded to the GitHub pages dedicated to ESP Launchpad. In general this action simplifies the process of building and publishing binary files for ESP-based projects, making it easier for developers to share their work and showcase examples.

It consists of:
- action.yml:
  - Action itself, it fetches the Python script, builds the examples with idf-build-apps using the configuration defined in `.idf_build_apps.toml`, then uses the Python script responsible for mergeing the binaries and creating config.toml which are uploaded to the github pages
  - Inputs: github token (With privilidges to the github repo in which this action is being used), idf_version (Used to build the binaries), parallel_count (Number of parallel builds), parallel_index (Index of parallel build)

- generateFiles.py:
  -  Python script responsible for processing the output json of idf-build-apps build, merging the binaries and creating config.toml for ESP-Launchpad.

XDanielPaul committed:
Release v1.0

XDanielPaul commented:
@kumekay @espzav PTAL

espzav commented on action.yml:
```diff
@@ -0,0 +1,57 @@
+        curl \
+          -H "Accept: application/vnd.github.v3.raw" \
+          -o generateFiles.py \
+          -L https://api.github.com/repos/XDanielPaul/idf-examples-launchpad-ci-action/contents/generateFiles.py?ref=v1.0
```
This path should be changed to new repository destination.

kumekay replied:
This URL must not be static, it should be the same version as the action

kumekay replied:
Maybe it is better to create a separate Dockerfile-based action in this repo with this file and use it as a step after idf-build-apps

XDanielPaul replied:
I removed the download of the script entirely and found out, that I can specify path to the script with `${GITHUB_ACTION_PATH}`

espzav reviewed (commented):
@XDanielPaul Good work! Thank you! I haven't got too much knowledge with CI and python, I left only one comment.

kumekay reviewed (changes requested):
@XDanielPaul Please take some time to choose a proper way for versioning of your action. Typical ways are described here: https://docs.github.com/en/actions/creating-actions/about-custom-actions#good-practices-for-release-management

For most of our actions we use branches: https://docs.github.com/en/actions/creating-actions/about-custom-actions#using-branches-for-release-management I.E. default branch is not master, but `v1`

tomassebestik commented:
Thank you @XDanielPaul for this tool. The code and docs are very clear, readable an nicely commented, I really like it.

LGTM, just few nitpicks:
- please ensure all your files end up with empty line (those ⛔ signs in some or your files diff); good idea is set your editor to do this automatically
- Danger is failing, because one of the checks expects the target branch should be `master`/`main` - I'm not sure why you are targeting release branch here.

radimkarnis commented on generateFiles.py:
```diff
@@ -0,0 +1,147 @@
+            for board in app['boards']:
+                kit = board[0]
+                target = board[1]
+                cmd = ['esptool.py', '--chip', target, 'merge_bin', '-o', app['name'] + '-' + kit + '-' + target + '-' + idf_version + '.bin', '@flash_args']
```
nitpick: you are joining strings here and using f-strings later (`f'{app["name"]}-{kit}-{target}-{idf_version}.bin'`). I would recommend using f-strings everywhere.

XDanielPaul replied:
@radimkarnis 

Thanks you for taking a look, I fixed it!

XDanielPaul commented:
> Thank you @XDanielPaul for this tool. The code and docs are very clear, readable an nicely commented, I really like it.
> 
> LGTM, just few nitpicks:
> 
> * please ensure all your files end up with empty line (those ⛔ signs in some or your files diff); good idea is set your editor to do this automatically
> * Danger is failing, because one of the checks expects the target branch should be `master`/`main` - I'm not sure why you are targeting release branch here.

Thank you for taking a look! I have addressed the issues

XDanielPaul commented:
From my point of view this PR is ready to be merged, please take a look

tomassebestik commented:
> From my point of view this PR is ready to be merged, please take a look

One more thing ... can you as part of this PR please update `.gitlab/workflows/danger.yaml` based on this example (you can paste it there as is): https://github.com/espressif/github-actions/tree/master/danger_pr_review#example-workflow-yaml-file

(after this update you should see Danger output here).

Many thanks.

XDanielPaul commented:
> > From my point of view this PR is ready to be merged, please take a look
> 
> One more thing ... can you as part of this PR please update `.gitlab/workflows/danger.yaml` based on this example (you can paste it there as is): https://github.com/espressif/github-actions/tree/master/danger_pr_review#example-workflow-yaml-file
> 
> (after this update you should see Danger output here).
> 
> Many thanks.

I have added it separately in order for the workflow to take an effect, because of the context of `pull_request_target`.  As mentioned [here](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target) the `pull_request_target` event "runs in the context of the base of the pull request, rather than in the context of the merge commit, as the pull_request event does."

tomassebestik commented:
> > > From my point of view this PR is ready to be merged, please take a look
> > 
> > 
> > One more thing ... can you as part of this PR please update `.gitlab/workflows/danger.yaml` based on this example (you can paste it there as is): https://github.com/espressif/github-actions/tree/master/danger_pr_review#example-workflow-yaml-file
> > (after this update you should see Danger output here).
> > Many thanks.
> 
> I have added it separately in order for the workflow to take an effect, because of the context of `pull_request_target`. As mentioned [here](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request_target) the `pull_request_target` event "runs in the context of the base of the pull request, rather than in the context of the merge commit, as the pull_request event does."

@XDanielPaul Well, it's not still running as should ... I suspect it's due to changing/targeting a different default branch after forking (the default branch for this project is set to `release/v1`) . Btw. If we don't use `master` at all, should'n be removed to avoid confusion?

The issue is probably related to the fact, with your fork you are going to `espressif:release/v1 from XDanielPaul:master` (you forked from `master`, then changed default branch and targeting back `release/v1`) and Danger (github-action bot) have some troubles with context.


Note: You don't have to fork at all, because you are listed in the repository as member. You can create a new branch directly in the repository and open a pull request directly :)

I did a quick test with fork with my test account and that works. Maybe that works because my fork goes to `espressif:release/v1 from tomassebestik-test:release/v1`
- https://github.com/espressif/idf-examples-launchpad-ci-action/pull/2

This is not a blocker for your MR ... just trying to figure out the issue core with Danger action.

tomassebestik reviewed (approved):
LGTM .... thank you @XDanielPaul, great job!

leeebo commented:
@XDanielPaul Great work! we'll use it in our projects. 

Would you consider supporting publishing the binaries to other storage servers like Amazon S3? Some countries are slow to access GitHub downloads 😢

hfudev commented on README.md:
````diff
@@ -0,0 +1,112 @@
+# Configuration file for the build process 
+config = "sdkconfig.defaults"
+```
+**Do not overwrite** `collect-app-info` **and** `build-dir` **options!**
````
Maybe also note why can't overwrite these two options?

hfudev commented on generateFiles.py:
```diff
@@ -0,0 +1,147 @@
+
+# Merge binaries for each app
+def merge_binaries(apps):
+    subprocess.run(['mkdir', 'binaries'])
```
`os.makedirs` would be better.

hfudev commented on generateFiles.py:
```diff
@@ -0,0 +1,147 @@
+                cwd = f'{app.get("app_dir")}/build_{kit}'
+                subprocess.run(cmd, cwd=cwd)
+                print(f'Merged binaries for {app["name"]}-{kit}-{target}-{idf_version}.bin')
+                subprocess.run(['mv', f'{cwd}/{app["name"]}-{kit}-{target}-{idf_version}.bin', 'binaries'])
```
here as well. `shutil.move` will be better, since is os independent.

hfudev reviewed (commented):
LGTM!

XDanielPaul commented:
> @XDanielPaul Great work! we'll use it in our projects.
> 
> Would you consider supporting publishing the binaries to other storage servers like Amazon S3? Some countries are slow to access GitHub downloads 😢

@leeebo I think it should be possible to publish binaries to the S3 through some 3rd party Github Actions, the github pages deployment is not mandatory, only recommended, because of CORS (ESP-Launchpad is hosted under `github.io` domain). So you will have to add CORS headers to your S3.

As mentioned in the note of 3rd step, the binaries and config.toml are in the `./binaries` directory after they are built, so you can use any method of transfering these files anywhere in your github workflow, you do not need to use github pages deplyoment.

XDanielPaul commented:
@hfudev Thank you for taking a look!
I have applied your suggestions
